timeline_shifts_visualization_experiment/02_parametric_vs_aligned_umap.py

```diff
@@ -0,0 +1,389 @@
+# -
+
+# # Parametric UMAP
+
+# ## Unaligned Models
+
+unaligned_word_models = list(
+    Path("../multi_model_experiment/output/models").rglob("*model")
+)
+unaligned_word_models = sorted(unaligned_word_models, key=lambda x: x.stem)
+unaligned_word_model_filter = list(
+    filter(lambda x: "_0" in x.stem, unaligned_word_models)
+)
+
+word_freq_count_cutoff = 5
+odd_year_subset = unaligned_word_model_filter[15:-1:2]
+
+training_unaligned_word_model_map = dict()
+for word_file in tqdm.tqdm(odd_year_subset):
+    model = Word2Vec.load(str(word_file)).wv
+    training_unaligned_word_model_map[word_file.stem.split("_")[0]] = dict(
+        model=model,
+        cutoff_index=min(
+            map(
+                lambda x: 999999
+                if model.get_vecattr(x[1], "count") > word_freq_count_cutoff
+                else x[0],
+                enumerate(model.index_to_key),
+            )
+        ),
+    )
+
+words_to_visualize = []
+token_character_list = []
+year_labels_list = []
+
+for year in tqdm.tqdm(training_unaligned_word_model_map):
+    model = training_unaligned_word_model_map[year]["model"]
+    word_subset_matrix = model[
+        model.index_to_key[: training_unaligned_word_model_map[year]["cutoff_index"]]
+    ]
+    print((year, word_subset_matrix.shape))
+    words_to_visualize.append(word_subset_matrix)
+    token_character_list += list(
+        map(
+            lambda x: re.escape(x),
+            model.index_to_key[
+                : training_unaligned_word_model_map[year]["cutoff_index"]
+            ],
+        )
+    )
+    year_labels_list += [year] * len(
+        model.index_to_key[: training_unaligned_word_model_map[year]["cutoff_index"]]
+    )
+
+training_unaligned_words = np.vstack(words_to_visualize)
+training_unaligned_words.shape
+
+umap_model_path = Path("output/unaligned_parametric_umap_model")
+if not umap_model_path.exists():
+    unaligned_parametric_model = ParametricUMAP(
+        verbose=True,
+        metric="cosine",
+        random_state=101,
+        low_memory=True,
+        n_neighbors=25,
+        min_dist=0.99,
+        n_training_epochs=1,
+    )
+    umap_embeddings = unaligned_parametric_model.fit_transform(training_unaligned_words)
+    unaligned_parametric_model.save(str(umap_model_path))
+else:
+    unaligned_parametric_model = load_ParametricUMAP(umap_model_path)
+    umap_embeddings = unaligned_parametric_model.transform(training_unaligned_words)
+
+train_plot_df = (
+    pd.DataFrame(umap_embeddings, columns=["umap1", "umap2"])
+    >> ply.define(year=year_labels_list, tok=token_character_list)
+    >> ply.call(".dropna")
+    >> ply.define(umap1="umap1.astype(float)")
+)
+train_plot_df >> ply.slice_rows(10)
+
+g = (
+    p9.ggplot(
+        train_plot_df >> ply.define(year="pd.Categorical(year)"),
+        p9.aes(x="umap1", y="umap2", fill="year"),
+    )
+    + p9.geom_point(alpha=0.2)
+    + p9.labs(title="Parameter Sweep of UMAP 2011-2020 (odd years)")
```

Comment:
should the title be "Parameter Sweep of UMAP 2015-2020 (odd years)" ?

Reply:
Good catch. I need to update that!